cpu/atmega_common/periph/rtc.c

```diff
@@ -0,0 +1,241 @@
+{
+    time_t time_secs;
+
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    time_secs = rtc_state.time + (TCNT2 >> 5);
+
+    /* Convert from seconds since the epoch */
+    gmtime_r(&time_secs, time);
+
+    DEBUG("RTC get time: %" PRIu32 " seconds\n", time_secs);
+
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    time_t now, diff;
+
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2B);
+    rtc_state.alarm_cb = NULL;
+
+    /* Make sure it is safe to read TCNT2, in case we just woke up, and */
+    /* safe to write OCR2B (in case it was busy) */
+    DEBUG("RTC sleeps until safe read TCNT2 and to write OCR2B\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    now = rtc_state.time + (TCNT2 >> 5);
+
+    /* Set alarm time */
+    rtc_state.alarm = mk_gmtime(time);
+
+    if (rtc_state.alarm < now) {
+        DEBUG("RTC alarm set in the past. Time: %" PRIu32 " seconds, alarm: %"
+              PRIu32 "\n", now, rtc_state.alarm);
+        cb(arg);
+        return 0;
+    }
+
+    /* Calculate time to wait, based on TCNT2 = 0 */
+    diff = rtc_state.alarm - rtc_state.time;
+
+    /* Prepare the counter for sub 8-second precision */
+    OCR2B = (uint8_t)(rtc_state.alarm << 5);
+
+    DEBUG("RTC set alarm: %" PRIu32 " seconds, OCR2B: %" PRIu8 "\n",
+          rtc_state.alarm, OCR2B);
+
+    /* Interrupt safe order of assignment */
+    rtc_state.alarm_arg = arg;
+    rtc_state.alarm_cb = cb;
+
+    /* Enable irq only if alarm is in the 8s period before it overflows */
+    if (diff < 8) {
+        /* Clear interrupt flag */
+        TIFR2 = (1 << OCF2B);
+
+        /* Enable interrupt */
+        TIMSK2 |= (1 << OCIE2B);
+
+        DEBUG("RTT alarm interrupt active\n");
+    }
+    else {
+        DEBUG("RTT alarm interrupt not active\n");
+    }
+
+    return 0;
+}
+
+int rtc_get_alarm(struct tm *time)
+{
+    /* Convert from seconds since the epoch */
+    /* Note: assignment is to discard volatile */
+    time_t alarm = rtc_state.alarm;
+    gmtime_r(&alarm, time);
+
+    DEBUG("RTC get alarm: %" PRIu32 " seconds\n", alarm);
+
+    return 0;
+}
+
+void rtc_clear_alarm(void)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2B);
+
+    /* Clear interrupt flag */
+    TIFR2 = (1 << OCF2B);
+
+    /* Interrupt safe order of assignment */
+    rtc_state.alarm_cb = NULL;
+    rtc_state.alarm_arg = NULL;
+}
+
+void rtc_poweron(void)
+{
+    rtt_poweron();
+}
+
+void rtc_poweroff(void)
+{
+    rtt_poweroff();
+}
+
+void atmega_rtc_incr(void)
+{
+    rtc_state.time += 8;
+
+    /* If alarm not set, nothing else to do */
+    if (rtc_state.alarm_cb == NULL) {
+        return;
+    }
+
+    /* Check to see if alarm was missed */
+    if (rtc_state.alarm <= rtc_state.time) {
+        rtc_state.alarm_cb(rtc_state.alarm_arg);
+        rtc_clear_alarm();
```

Comment:
Do not clear after the callback, it might have set a new alarm.

Reply:
Fixed.